lib/engine/depot.rb

```diff
@@ -70,30 +70,28 @@ def max_depot_price
       train.variants.map { |_, v| v[:price] }.max
     end
 
-    def unshift_train(train)
-      train.owner = self
-      @upcoming.unshift(train)
-      @depot_trains = nil
-    end
-
     def remove_train(train)
-      @upcoming.delete(train)
+      delete_from_upcoming(train)
       @discarded.delete(train)
       @depot_trains = nil
     end
 
     def forget_train(train)
       @trains.delete(train)
-      @upcoming.delete(train)
-      @discarded.delete(train)
-      @depot_trains = nil
+      remove_train(train)
+    end
+
+    def delete_from_upcoming(train)
+      index = @upcoming.index(train)
+      deleted = @upcoming.delete(train)
+      return if !index || !deleted&.unlimited
+
+      insert_train(train.clone_unlimited, index)
```

Comment:
Is index correct here? If you have 3 unlimited trains (due to an event on 3rd train), when the first unlimited train is removed, will this insert a new train at that position and thus we'll never get to the 3rd train?

Reply:
That’s a good call. I added this to fix an 1889 game where a D was bought before the second 6, wasn’t thinking about events at that moment.

So the answer I think is to get the index of the last train with the same name as the removed one before it is actually removed.

Reply:
Something like that would work. It may be simpler to check if index + 1 has the same name.

`return if !index || !deleted&.unlimited || @upcoming[index + 1]&.name == train.name`

Reply:
Right, that logic will need to be refactored. Something like `remove_train(train, export = true)`.